### PR TITLE
battle (3rd take)

### PR DESCRIPTION
**Pull request by jiangzhengwenjz**

### CODE REVIEW

**Comment by jiangzhengwenjz** [comment]
Ready for review.

File: asm/battle_controller_pokedude.s [hunk 1]
```diff
@@ -2862,7 +2862,7 @@ sub_81577B4: @ 81577B4
 	ldr r0, _081577E4 @ =gActiveBattler
 	ldrb r0, [r0]
 	lsls r2, r0, 9
-	ldr r1, _081577E8 @ =gUnknown_2022BC7
+	ldr r1, _081577E8 @ =(gBattleBufferA + 3)
```
Comment (PikalaxALT) [inline asm/battle_controller_pokedude.s]: Parentheses are not necessary here or in the pools.

File: data/data_83F5738.s [hunk 2]
```diff
@@ -345,7 +345,28 @@ gUnknown_83FE80C:: @ 83FE80C
...
+	.incbin "baserom.gba", 0x3FE83C, 0xA
+
+gText_Ice:: @ 83FE846
+	.incbin "baserom.gba", 0x3FE846, 0x4
+
+gText_Confusion:: @ 83FE84A
+	.incbin "baserom.gba", 0x3FE84A, 0xA
+	
+gText_Love:: @ 83FE854
+	.incbin "baserom.gba", 0x3FE854, 0x5
```
Comment (PikalaxALT) [inline data/data_83F5738.s]:
```armasm
gText_HighlightRed_Left::
	.string "{HIGHLIGHT RED}Left: $"

gText_HighlightRed::
	.string "{HIGHLIGHT RED}$"

gText_Sleep::
	.string "sleep$"

gText_Poison::
	.string "poison$"

gText_Burn::
	.string "burn$"

gText_Paralysis::
	.string "paralysis$"

gText_Ice::
	.string "ice$"

gText_Confusion::
	.string "confusion$"

gText_Love::
	.string "love$"
```

File: include/constants/battle_anim.h [hunk 3]
```diff
@@ -377,6 +377,7 @@
 #define B_ANIM_SWITCH_OUT_PLAYER_MON    0x1
 #define B_ANIM_SWITCH_OUT_OPPONENT_MON  0x2
 #define B_ANIM_BALL_THROW               0x3
+#define B_ANIM_SP_4                     0x4 // this one is used in battle_controller_oak_old_man
 #define B_ANIM_SAFARI_BALL_THROW        0x4
```
Comment (PikalaxALT) [inline include/constants/battle_anim.h]: Are they two separate anims with the same index?

File: include/constants/battle_setup.h [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+#define TRAINER_BATTLE_SINGLE                           0
+#define TRAINER_BATTLE_CONTINUE_SCRIPT_NO_MUSIC         1
+#define TRAINER_BATTLE_CONTINUE_SCRIPT                  2
+#define TRAINER_BATTLE_SINGLE_NO_INTRO_TEXT             3
+#define TRAINER_BATTLE_DOUBLE                           4
+#define TRAINER_BATTLE_REMATCH                          5
+#define TRAINER_BATTLE_CONTINUE_SCRIPT_DOUBLE           6
+#define TRAINER_BATTLE_REMATCH_DOUBLE                   7
+#define TRAINER_BATTLE_CONTINUE_SCRIPT_DOUBLE_NO_MUSIC  8
+#define TRAINER_BATTLE_TUTORIAL                         9
```
Comment (PikalaxALT) [inline include/constants/battle_setup.h]: Inject these into the event scripts please.

File: include/constants/map_types.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    MAP_BATTLE_SCENE_MAGMA,        // 2
+    MAP_BATTLE_SCENE_AQUA,         // 3
+    MAP_BATTLE_SCENE_SIDNEY,       // 4
+    MAP_BATTLE_SCENE_PHOEBE,       // 5
+    MAP_BATTLE_SCENE_GLACIA,       // 6
+    MAP_BATTLE_SCENE_DRAKE,        // 7
+    MAP_BATTLE_SCENE_BATTLE_TOWER, // 8
+};
+
+#endif  // GUARD_CONSTANTS_MAP_TYPES_H
```
Comment (PikalaxALT) [inline include/constants/map_types.h]: Missing trailing newline

File: include/constants/map_types.h [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+#define MAP_TYPE_CITY        2
+#define MAP_TYPE_ROUTE       3
+#define MAP_TYPE_UNDERGROUND 4
+#define MAP_TYPE_UNDERWATER  5
+#define MAP_TYPE_OCEAN_ROUTE 6
+#define MAP_TYPE_7           7
+#define MAP_TYPE_INDOOR      8
+#define MAP_TYPE_SECRET_BASE 9
+
+enum
```
Comment (PikalaxALT) [inline include/constants/map_types.h]: If these ever need to be included in an asm file, this enum must first be converted into `#define` tokens.

File: include/constants/trainers.h [hunk 7]
```diff
@@ -165,6 +165,8 @@
 #define TRAINER_PIC_LADY                  146
 #define TRAINER_PIC_PAINTER               147
 
+#define TRAINER_SECRET_BASE                 1024
```
Comment (PikalaxALT) [inline include/constants/trainers.h]: oh dear LUL

File: src/fame_checker.c [hunk 8]
```diff
@@ -878,7 +878,7 @@ static void FC_DestroyWindow(u8 windowId)
 
 static u8 AdjustGiovanniIndexIfBeatenInGym(u8 a0)
 {
-    if (HasTrainerAlreadyBeenFought(0x15e) == TRUE)
+    if (HasTrainerBeenFought(0x15e) == TRUE)
```
Comment (PikalaxALT) [inline src/fame_checker.c]: Use the constant from constants/trainers.h

File: src/item.c [hunk 9]
```diff
@@ -73,7 +73,7 @@ void CopyItemName(u16 itemId, u8 * dest)
 {
     if (itemId == ITEM_ENIGMA_BERRY)
     {
-        StringCopy(dest, sub_809C8A0(43)->name);
+        StringCopy(dest, GetBerryInfo(43)->name);
```
Comment (PikalaxALT) [inline src/item.c]: Use the constant from constants/items.h

File: src/unk_8159F40.c [hunk 10]
```diff
@@ -222,7 +222,7 @@ void sub_815A1F8(const struct QuestLog * questLog, const struct MapObjectTemplat
         gMapObjects[i].mapobj_bit_24 = questLogMapObjects[i].mapobj_bit_24;
         gMapObjects[i].mapobj_bit_25 = questLogMapObjects[i].mapobj_bit_25;
         gMapObjects[i].mapobj_bit_26 = questLogMapObjects[i].mapobj_bit_26;
-        gMapObjects[i].mapobj_unk_18 = questLogMapObjects[i].mapobj_unk_18;
+        gMapObjects[i].facingDirection = questLogMapObjects[i].mapobj_unk_18;
```
Comment (PikalaxALT) [inline src/unk_8159F40.c]: Think you can get this function matching?

File: src/vs_seeker.c [hunk 11]
```diff
@@ -1027,7 +1027,7 @@ static u8 GetVsSeekerResponseInArea(const VsSeekerData * a0)
                 "\tadds r0, r5\n"
                 "\tldrh r0, [r0, 0x4]\n"
                 "\tmov r8, r0\n"
-                "\tbl HasTrainerAlreadyBeenFought\n"
+                "\tbl HasTrainerBeenFought\n"
```
Comment (PikalaxALT) [inline src/vs_seeker.c]: Think you can match this?

File: include/constants/map_types.h (hunk 6 above)
Comment (jiangzhengwenjz) [reply include/constants/map_types.h]: my bad, should not have copied this one

File: src/battle_main.c [hunk 12]
```diff
@@ -0,0 +1,4414 @@
...
+u8 gLeveledUpInBattle;
+void (*gBattlerControllerFuncs[MAX_BATTLERS_COUNT])(void);
+u8 gHealthboxSpriteIds[MAX_BATTLERS_COUNT];
+u8 gMultiUsePlayerCursor;
+u8 gNumberOfMovesToChoose;
+u8 gUnknown_3004FFC[MAX_BATTLERS_COUNT];
+
+static const struct ScanlineEffectParams sIntroScanlineParams16Bit =
+{
+    (void *)REG_ADDR_BG3HOFS, SCANLINE_EFFECT_DMACNT_16BIT, 1
```
Comment (PikalaxALT) [inline src/battle_main.c]: `&REG_BG3HOFS`

File: src/battle_main.c [hunk 13]
```diff
@@ -0,0 +1,4414 @@
...
+
+static void sub_8011BB0(void)
+{
+    s32 i;
+
+    switch (gBattleCommunication[MULTIUSE_STATE])
+    {
+    case 0:
+        ShowBg(0);
+        Sh
```
Comment (PikalaxALT) [inline src/battle_main.c]: Return value:
0 = Can run
1 = Can't run,  consume turn
2 = Can't run, don't consume turn

File: src/battle_setup.c [hunk 14]
```diff
@@ -0,0 +1,1059 @@
...
+    u8 minPartyCount;
+    u8 transitionType;
+    u8 enemyLevel;
+    u8 playerLevel;
+
+    if (gTrainerBattleOpponent_A == TRAINER_SECRET_BASE)
+        return B_TRANSITION_BLUE;
+    if (gTrainers[gTrainerBattleOpponent_A].trainerClass == CLASS_ELITE_FOUR_2)
+    {
+        if (gTrainerBattleOpponent_A == 410 || gTrainerBattleOpponent_A == 735) // TODO: replace these with enums
```
Comment (PikalaxALT) [inline src/battle_setup.c]: TRAINER_ELITE_FOUR_LORELEI || TRAINER_ELITE_FOUR_LORELEI_2

File: src/battle_setup.c [hunk 15]
```diff
@@ -0,0 +1,1059 @@
...
+void ScrSpecial_ShowTrainerNonBattlingSpeech(void)
+{
+    ShowFieldMessage(GetTrainerCantBattleSpeech());
+}
+
+void PlayTrainerEncounterMusic(void)
+{
+    u16 music;
+
+    if ((u8)(gUnknown_203ADFA - 2) > 1
```
Comment (PikalaxALT) [inline src/battle_setup.c]: `if (gUnknown_203ADFA != 2 && gUnknown_203ADFA != 3`

**Pull request merged**